apps_download/models/product_product.py

```diff
@@ -0,0 +1,107 @@
+# -*- coding: utf-8 -*-
+
+import os
+import tempfile
+import shutil
+import logging
+import base64
+import subprocess
+import time
+from odoo.exceptions import ValidationError
+from odoo import models, fields, api, _
+_logger = logging.getLogger(__name__)
+
+
+class ProductProduct(models.Model):
+    _inherit = 'product.product'
+
+    dependent_product_ids = fields.Many2many(
+        'product.product', 'prto_validateoduct_dependent_rel',
+        'src_id', 'dest_id', string='Dependent Products'
+    )
+    module_path = fields.Char('Module Path')
+
+    @api.constrains('dependent_product_ids')
+    def check_dependent_recursion(self):
+        for product in self:
+            def child_dependency_check(product_dependent_ids, children):
+                res = self.env['product.product']
+                for child in children:
+                    if not child.dependent_product_ids:
+                        continue
+                    if child in product_dependent_ids:
+                        raise ValidationError(
+                            _('Error: You cannot create recursive dependency.')
+                        )
+                    product_dependent_ids += child
+                    child_dependency_check(product_dependent_ids,
+                                           child.dependent_product_ids)
+                return res
+
+            child_dependency_check(product, product.dependent_product_ids)
+
+    @api.multi
+    def create_dependency_list(self):
+        ret_val = {}
+
+        def child_dependency(children):
```

Comment:
IMO this should be a first-class method and not nested inside of another method. Nesting makes it impossible to change in child modules, and is also near impossible to test properly.

Reply:
@agenterp  Can you please take care for this.